### PR TITLE
MULE-17674: Only allow MuleSession instances when deserializing corresponding header

### PR DESCRIPTION
**Pull request by eze210**

### CODE REVIEW

File: core/src/main/java/org/mule/session/AbstractSessionHandler.java [hunk 1]
```diff
@@ -45,6 +47,18 @@
...
 
+    private ObjectSerializer setSerializerClass(ObjectSerializer serializer, Class clazz)
+    {
+        if (serializer instanceof ClassSpecificObjectSerializer)
+        {
+            return serializer;
+        }
+        else
+        {
+            return new ClassSpecificObjectSerializer(serializer, clazz);
```
Comment (aiannucci) [inline core/src/main/java/org/mule/session/AbstractSessionHandler.java]: Will this create a new instance every time? Is this really needed?

File: core/src/test/java/org/mule/MuleSessionHandlerTestCase.java [hunk 2]
```diff
@@ -119,6 +124,20 @@ public void testSessionProperties() throws Exception
         assertEquals(list, obj);
     }
 
+    @Test
+    public void testDontDeserializeOtherThanMuleSession() throws MuleException
+    {
+        DefaultMuleMessage message = new DefaultMuleMessage("Test Message", muleContext);
+        SessionHandler handler = new SerializeAndEncodeSessionHandler();
+
+        String encodedSet = new String(Base64.encode(SerializationUtils.serialize(new HashSet<String>())));
```
Comment (fsgonz) [inline core/src/test/java/org/mule/MuleSessionHandlerTestCase.java]: static import Base64.encode and SerializationUtils.serialize

File: core/src/main/java/org/mule/serialization/internal/ClassSpecificObjectSerializer.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        this.classToAccept = classToAccept;
+    }
+
+    @Override
+    protected <T> T doDeserialize(InputStream inputStream, ClassLoader classLoader) throws Exception
+    {
+        checkArgument(inputStream != null, "Cannot deserialize a null stream");
+        checkArgument(classLoader != null, "Cannot deserialize with a null classloader");
+
+        return (T) SerializationUtils.deserialize(inputStream, classLoader, muleContext, new ObjectInputStreamProvider()
```
Comment (fsgonz) [inline core/src/main/java/org/mule/serialization/internal/ClassSpecificObjectSerializer.java]: static import.

File: core/src/main/java/org/mule/serialization/internal/ClassSpecificObjectSerializer.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+         * @param classLoader the ClassLoader from which classes should be loaded
+         * @param inputStream the InputStream to work on
+         * @throws IOException              in case of an I/O error
+         * @throws StreamCorruptedException if the stream is corrupted
+         */
+        public ClassSpecificObjectInputStream(ClassLoader classLoader, InputStream inputStream, Class classToAccept) throws IOException, StreamCorruptedException
+        {
+            super(classLoader, inputStream);
+            this.classToAccept = classToAccept;
+            this.firstClass = true;
```
Comment (fsgonz) [inline core/src/main/java/org/mule/serialization/internal/ClassSpecificObjectSerializer.java]: set the true value in the field declaration.

File: core/src/main/java/org/mule/serialization/internal/ClassSpecificObjectSerializer.java (hunk 3 above)
Comment (eze210) [reply core/src/main/java/org/mule/serialization/internal/ClassSpecificObjectSerializer.java]: The base class (AbstractObjectSerializer) already declares a deserialize() method

Comment (fsgonz) [reply core/src/main/java/org/mule/serialization/internal/ClassSpecificObjectSerializer.java]: You're right.

**Pull request merged**